### PR TITLE
Factor out l3d, g3d, lnd loading into Component

### PR DESCRIPTION
**Pull request by bwrsandman**
I'm looking for feedback on this and it needs a few more features such as writing files and extracting meshes, textures. Maybe different options are better? Subcomands?

* Add a component libraries called l3d, g3d, lnd that can read and write files
* Use libraries when loading l3d, g3d, lnd file from filename.
* Add apps: l3dtool, g3dtool, lndtool inspired by https://github.com/OpenMW/openmw/tree/master/apps/bsatool

## l3dtool
```
Inspect and extract files from LionHead L3D files.
Usage:
  l3dtool [OPTION...]

  -h, --help                    Display this help message.
  -H, --header arg              Print Header Contents.
  -m, --mesh-header arg         Print Mesh Headers.
  -s, --skins arg               Print Skins.
  -p, --points arg              Print Points.
  -P, --primitive-header arg    Print Primitive Headers.
  -b, --bones arg               Print Bones.
  -V, --vertices arg            Print Vertices.
  -I, --indices arg             Print Indices.
  -L, --look-up-tables arg      Print Look Up Table Data.
  -B, --vertex-blend-values arg
                                Print Vertex Blend Values.
```

```
$ l3dtool -H Data/WeatherSystem/sky.l3d      

file: Data/WeatherSystem/sky.l3d
magic: L3D0
flags: 0x00022000
size: 7064
submesh count: 1
submesh start offset: 0x0000004C
unknown 32 bytes: 0000000000000000000000000000000000000000000000000000000000000000
unknown offset: 0xFFFFFFFF
skin count: 0
skin start offset: 0x00001B98
point count: 0
point start offset: 0xFFFFFFFF
extra start offset: 0xFFFFFFFF
```

```
$ 3dtool -m Data/Misc/coffre.l3d
file: Data/Misc/coffre.l3d
mesh #1
flags: 0xE0000801
primitive count: 8
primitives start offset: 0x000002BC
bone count: 10
bones start offset: 0x00000064
```

```
$ l3dtool -P Data/CreatureMesh/C_Cow_Good.l3d
file: CreatureMesh/C_Cow_Good.l3d
primitive #1
unknown_1: 0x00000002
unknown_2: 0x00000000
skinID: 0xFBC68792
unknown_2: 0xFFFF0000
vertex count: 269
vertex start offset: 0x00000BE0
triangle count: 194
triangle start offset: 0x00002D80
bone vertex look-up table size: 33
bone vertex look-up table start offset: 0x00003514
vertex blend count: 97
vertex blend start offset: 0x0000320C

primitive #2
unknown_1: 0x00000002
unknown_2: 0x00000000
skinID: 0xFA795085
unknown_2: 0xFFFFC39B
vertex count: 461
vertex start offset: 0x000035C8
triangle count: 400
triangle start offset: 0x00006F68
bone vertex look-up table size: 30
bone vertex look-up table start offset: 0x00007EF8
vertex blend count: 198
vertex blend start offset: 0x000078C8

primitive #3
unknown_1: 0x00000002
unknown_2: 0x00000000
skinID: 0xFB8EBD24
unknown_2: 0xFFFFFFFF
vertex count: 355
vertex start offset: 0x00007FA0
triangle count: 370
triangle start offset: 0x0000AC00
bone vertex look-up table size: 24
bone vertex look-up table start offset: 0x0000B66C
vertex blend count: 56
vertex blend start offset: 0x0000B4AC
```

```
$ l3dtool --skins Data/CreatureMesh/A_Bear_Boned_base.L3D

file: Data/CreatureMesh/A_Bear_Boned_base.L3D
skin id: 0xF661404F
data:

```

Screenshot from 2019-11-11 20-27-54

```
$ l3dtool -I Data/CreatureMesh/C_Cow_Good.l3d
file: Data/CreatureMesh/C_Cow_Good.l3d
  11    0    1   13   12   14   16    0   11   23   22   24
  22   25   24   23   26   40   22   23   41   41   42   22
  43   22   42   22   43   25   41   23   40   42   41   44
  40   44   41   45   42   44   42   45   43   44   40   46
  46   47   44   45   44   48   49   48   44   49   47   50
  50    7   49   48   49    8    7    8   49    9    7   50
   7   51   18    8   18   19   18    8    7   55   54   66
  67   66   54   83   59   84   85   84   59   84   85   91
  92   91   85   93   92   94   94  102   93  102  103  107
 108  107  103  107  108  109  110  109  108  115  114  116
# ...
```

```
$ l3dtool -V Data/CreatureMesh/C_Cow_Good.l3d
file: Data/CreatureMesh/C_Cow_Good.l3d
|     position     |   uv coord   |        normal        |
|------------------|--------------|----------------------|
|  2.3   1.0  -1.2 | 0.047  0.959 |  0.532  0.036 -0.846 |
|  2.6  -0.1  -1.0 | 0.071  0.963 |  0.607 -0.253 -0.754 |
| -3.9   0.2  -1.1 | 0.309  0.327 | -0.785 -0.394  0.478 |
| -0.0  -0.0  -0.4 | 0.450  0.290 | -0.785 -0.394  0.478 |
|  0.8   0.2   1.0 | 0.506  0.239 | -0.785 -0.394  0.478 |
|  0.3   0.9   0.2 | 0.471  0.268 | -0.785 -0.394  0.478 |
|  0.6   0.6  -1.1 | 0.459  0.309 | -0.785 -0.394  0.478 |
|  2.1   1.2   0.4 | 0.538  0.253 | -0.413  0.850  0.327 |
# ...
```

```
$ l3dtool -b Data/Misc/coffre.l3d
file: Data/Misc/coffre.l3d
bone #0
parent: None
first child: 1
right sibling: None
rotation:
     [ 1.000  0.000 -0.000]
     [ 0.000  0.000  1.000]
     [-0.000  1.000 -0.000]
position: (  0.0   0.4  -0.0)

bone #1
parent: 0
first child: None
right sibling: 2
rotation:
     [ 0.000  1.000  0.000]
     [-0.000 -0.000  1.000]
     [ 1.000  0.000  0.000]
position: ( -0.0   0.3   0.6)

bone #2
parent: 0
first child: None
right sibling: 3
rotation:
     [ 0.000  1.000  0.000]
     [-0.000 -0.000  1.000]
     [ 1.000  0.000  0.000]
position: ( -0.1   0.3   0.6)

bone #3
parent: 0
first child: None
right sibling: 4
rotation:
     [ 0.000  1.000  0.000]
     [-0.000 -0.000  1.000]
     [ 1.000  0.000  0.000]
position: ( -0.6   0.0   0.6)

bone #4
parent: 0
first child: None
right sibling: 5
rotation:
     [ 0.000  1.000  0.000]
     [-0.000 -0.000  1.000]
     [ 1.000  0.000  0.000]
position: ( -0.4  -0.4   0.6)

bone #5
parent: 0
first child: None
right sibling: 6
rotation:
     [ 0.000  1.000  0.000]
     [-0.000 -0.000  1.000]
     [ 1.000  0.000  0.000]
position: ( -0.0  -0.4   0.6)

bone #6
parent: 0
first child: None
right sibling: 7
rotation:
     [ 0.000  1.000  0.000]
     [-0.000 -0.000  1.000]
     [ 1.000  0.000  0.000]
position: (  0.0   0.3   0.6)

bone #7
parent: 0
first child: None
right sibling: 8
rotation:
     [ 0.000  1.000  0.000]
     [-0.000 -0.000  1.000]
     [ 1.000  0.000  0.000]
position: (  0.6  -0.0   0.6)

bone #8
parent: 0
first child: None
right sibling: 9
rotation:
     [ 0.000  1.000  0.000]
     [-0.000 -0.000  1.000]
     [ 1.000  0.000  0.000]
position: (  0.5  -0.4   0.6)

bone #9
parent: 0
first child: None
right sibling: None
rotation:
     [ 1.000  0.000  0.000]
     [ 0.000  1.000  0.000]
     [-0.000 -0.000  1.000]
position: (  0.0   0.6   0.5)
```
```
$ 3dtool -L Data/CreatureMesh/C_Cow_Good.l3d
file: Data/CreatureMesh/C_Cow_Good.l3d
       00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
     ,------------------------------------------------
0000 | 0B 00 22 00 04 00 26 00 07 00 23 00 0E 00 20 00
0010 | 04 00 06 00 0B 00 21 00 03 00 24 00 03 00 04 00
0020 | 09 21 00 20 00 1F 00 08 00 19 00 07 00 0B 00 0A
0030 | 00 24 00 06 00 0D 00 21 00 0C 00 22 00 06 00 73
0040 | 00 0C 00 07 00 04 00 08 00 00 00 0F 00 2A 00 02
0050 | 00 08 00 02 00 09 00
```
```
$ l3dtool -B Data/CreatureMesh/C_Cow_Good.l3d
file: Data/CreatureMesh/C_Cow_Good.l3d
       00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
     ,------------------------------------------------
0000 | 0B 00 22 00 04 00 26 00 07 00 23 00 0E 00 20 00
0010 | 04 00 06 00 0B 00 21 00 03 00 24 00 03 00 04 00
0020 | 09 21 00 20 00 1F 00 08 00 19 00 07 00 0B 00 0A
0030 | 00 24 00 06 00 0D 00 21 00 0C 00 22 00 06 00 73
0040 | 00 0C 00 07 00 04 00 08 00 00 00 0F 00 2A 00 02
0050 | 00 08 00 02 00 09 00
```

## g3dtool
```
$ g3dtool -h
Inspect and extract files from LionHead G3D files.
Usage:
  g3dtool [OPTION...] mesh-pack-files...

  -h, --help            Display this help message.
  -l, --list-blocks     List block statistics.
  -b, --view-bytes arg  View raw byte content of block.
  -i, --info-block      List INFO block statistics.
  -M, --meshes-block    List MESHES block statistics.
  -m, --mesh arg        List mesh statistics.
  -T, --texture-block   View texture block statistics.
  -t, --texture arg     View texture statistics.
  -e, --extract arg     Extract contents of a block (texture) to filename.
  -w, --write arg       Create Mesh Pack.
```

```
$ g3dtool -l Data/AllMeshes.g3d 
file: Data/AllMeshes.g3d
222 blocks
110 textures
626 meshes
1: name "1", size 32908
2: name "10", size 65676
3: name "11", size 65676
4: name "12", size 65676
5: name "13", size 65676
6: name "14", size 32908
7: name "15", size 32908
8: name "16", size 65676
9: name "17", size 32908
10: name "18", size 65676
11: name "19", size 65676
12: name "1a", size 32908
13: name "1b", size 65676
14: name "1c", size 32908
15: name "1d", size 65676
16: name "1e", size 32908
17: name "1f", size 65676
18: name "2", size 65676
19: name "20", size 32908
20: name "21", size 32908
21: name "22", size 65676
22: name "23", size 65676
23: name "24", size 65676
24: name "25", size 65676
25: name "26", size 65676
26: name "27", size 65676
27: name "28", size 32908
28: name "29", size 32908
29: name "2a", size 32908
30: name "2b", size 65676
31: name "2c", size 65676
32: name "2d", size 32908
33: name "2e", size 32908
34: name "2f", size 32908
35: name "3", size 32908
36: name "30", size 65676
37: name "31", size 65676
38: name "32", size 65676
39: name "33", size 32908
40: name "34", size 32908
41: name "35", size 32908
42: name "36", size 32908
43: name "37", size 65676
44: name "38", size 65676
45: name "39", size 32908
46: name "3a", size 32908
47: name "3b", size 65676
48: name "3c", size 32908
49: name "3d", size 32908
50: name "3e", size 32908
51: name "3f", size 32908
52: name "4", size 65676
53: name "40", size 65676
54: name "41", size 65676
55: name "42", size 65676
56: name "43", size 32908
57: name "44", size 32908
58: name "45", size 32908
59: name "46", size 65676
60: name "47", size 65676
61: name "48", size 65676
62: name "49", size 65676
63: name "4a", size 32908
64: name "4b", size 65676
65: name "4c", size 32908
66: name "4d", size 65676
67: name "4e", size 65676
68: name "4f", size 32908
69: name "5", size 65676
70: name "50", size 32908
71: name "51", size 32908
72: name "52", size 32908
73: name "53", size 32908
74: name "54", size 32908
75: name "55", size 32908
76: name "56", size 32908
77: name "57", size 32908
78: name "58", size 32908
79: name "59", size 32908
80: name "5a", size 32908
81: name "5b", size 32908
82: name "5c", size 32908
83: name "5d", size 65676
84: name "5e", size 65676
85: name "5f", size 32908
86: name "6", size 65676
87: name "60", size 65676
88: name "61", size 32908
89: name "62", size 65676
90: name "63", size 32908
91: name "64", size 65676
92: name "65", size 32908
93: name "66", size 65676
94: name "67", size 65676
95: name "68", size 65676
96: name "69", size 65676
97: name "6a", size 65676
98: name "6b", size 65676
99: name "6c", size 65676
100: name "6d", size 65676
101: name "6e", size 65676
102: name "7", size 65676
103: name "8", size 32908
104: name "9", size 65676
105: name "INFO", size 4100
106: name "LOW1", size 8332
107: name "LOW10", size 8332
108: name "LOW11", size 8332
109: name "LOW12", size 8332
110: name "LOW13", size 8332
111: name "LOW14", size 8332
112: name "LOW15", size 8332
113: name "LOW16", size 8332
114: name "LOW17", size 8332
115: name "LOW18", size 8332
116: name "LOW19", size 8332
117: name "LOW1a", size 8332
118: name "LOW1b", size 8332
119: name "LOW1c", size 8332
120: name "LOW1d", size 8332
121: name "LOW1e", size 8332
122: name "LOW1f", size 8332
123: name "LOW2", size 8332
124: name "LOW20", size 8332
125: name "LOW21", size 8332
126: name "LOW22", size 8332
127: name "LOW23", size 8332
128: name "LOW24", size 8332
129: name "LOW25", size 8332
130: name "LOW26", size 8332
131: name "LOW27", size 8332
132: name "LOW28", size 8332
133: name "LOW29", size 8332
134: name "LOW2a", size 8332
135: name "LOW2b", size 8332
136: name "LOW2c", size 8332
137: name "LOW2d", size 8332
138: name "LOW2e", size 8332
139: name "LOW2f", size 8332
140: name "LOW3", size 8332
141: name "LOW30", size 8332
142: name "LOW31", size 8332
143: name "LOW32", size 8332
144: name "LOW33", size 8332
145: name "LOW34", size 8332
146: name "LOW35", size 8332
147: name "LOW36", size 8332
148: name "LOW37", size 8332
149: name "LOW38", size 8332
150: name "LOW39", size 8332
151: name "LOW3a", size 8332
152: name "LOW3b", size 8332
153: name "LOW3c", size 8332
154: name "LOW3d", size 8332
155: name "LOW3e", size 8332
156: name "LOW3f", size 8332
157: name "LOW4", size 8332
158: name "LOW40", size 8332
159: name "LOW41", size 8332
160: name "LOW42", size 8332
161: name "LOW43", size 8332
162: name "LOW44", size 8332
163: name "LOW45", size 8332
164: name "LOW46", size 8332
165: name "LOW47", size 8332
166: name "LOW48", size 8332
167: name "LOW49", size 8332
168: name "LOW4a", size 8332
169: name "LOW4b", size 8332
170: name "LOW4c", size 8332
171: name "LOW4d", size 8332
172: name "LOW4e", size 8332
173: name "LOW4f", size 8332
174: name "LOW5", size 8332
175: name "LOW50", size 8332
176: name "LOW51", size 8332
177: name "LOW52", size 8332
178: name "LOW53", size 8332
179: name "LOW54", size 8332
180: name "LOW55", size 8332
181: name "LOW56", size 8332
182: name "LOW57", size 8332
183: name "LOW58", size 8332
184: name "LOW59", size 8332
185: name "LOW5a", size 8332
186: name "LOW5b", size 8332
187: name "LOW5c", size 8332
188: name "LOW5d", size 8332
189: name "LOW5e", size 8332
190: name "LOW5f", size 8332
191: name "LOW6", size 8332
192: name "LOW60", size 8332
193: name "LOW61", size 8332
194: name "LOW62", size 8332
195: name "LOW63", size 8332
196: name "LOW64", size 8332
197: name "LOW65", size 8332
198: name "LOW66", size 8332
199: name "LOW67", size 8332
200: name "LOW68", size 8332
201: name "LOW69", size 8332
202: name "LOW6a", size 8332
203: name "LOW6b", size 8332
204: name "LOW6c", size 8332
205: name "LOW6d", size 8332
206: name "LOW6e", size 8332
207: name "LOW7", size 8332
208: name "LOW8", size 8332
209: name "LOW9", size 8332
210: name "LOWa", size 8332
211: name "LOWb", size 8332
212: name "LOWc", size 8332
213: name "LOWd", size 8332
214: name "LOWe", size 8332
215: name "LOWf", size 8332
216: name "MESHES", size 6960410
217: name "a", size 32908
218: name "b", size 32908
219: name "c", size 65676
220: name "d", size 32908
221: name "e", size 32908
222: name "f", size 32908
```

## lndtool
```
$ lndtool -h
Inspect and extract files from LionHead LND files.
Usage:
  lndtool [OPTION...]

  -h, --help                    Display this help message.
  -H, --header arg              Print Header Contents.
  -l, --low-resolution-textures arg
                                Print Low Resolution Texture Contents.
  -b, --blocks arg              Print Block Contents.
  -c, --country arg             Print Country Contents.
  -m, --material arg            Print Material Contents.
  -x, --extra arg               Print Extra Content.
  -u, --unaccounted arg         Print Unaccounted bytes Content.
```

```
$ lndtool -H bw/Data/Landscape/Land1.lnd
file: Data/Landscape/Land1.lnd
block count: 101
block index look-up table:
       00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
     ,------------------------------------------------
0000 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0010 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0020 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0030 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0040 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0050 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0060 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0070 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0080 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0090 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00A0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00B0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00C0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00D0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00E0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01 02
00F0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0100 | 00 00 00 00 00 00 00 00 00 00 00 00 00 03 04 05
0110 | 06 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0120 | 00 00 00 00 00 00 00 00 00 00 00 00 07 08 09 0A
0130 | 0B 0C 0D 0E 00 00 00 00 00 00 00 00 00 00 00 00
0140 | 00 00 00 00 00 00 00 00 00 00 00 00 0F 10 11 12
0150 | 13 14 15 16 17 18 00 00 00 00 00 00 00 00 00 00
0160 | 00 00 00 00 00 00 00 00 00 00 00 00 00 19 1A 1B
0170 | 1C 1D 1E 1F 20 21 22 00 00 00 00 00 00 00 00 00
0180 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 23 24
0190 | 25 26 27 28 29 2A 2B 00 00 00 00 00 00 00 00 00
01A0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 2C 2D
01B0 | 2E 2F 30 31 32 33 34 00 00 00 00 00 00 00 00 00
01C0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 35 36 37
01D0 | 38 39 3A 3B 3C 3D 00 00 00 00 00 00 00 00 00 00
01E0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 3E 3F 40
01F0 | 41 42 43 44 45 46 00 00 00 00 00 00 00 00 00 00
0200 | 00 00 00 00 00 00 00 00 00 00 00 00 00 47 48 49
0210 | 4A 4B 4C 4D 4E 4F 00 00 00 00 00 00 00 00 00 00
0220 | 00 00 00 00 00 00 00 00 00 00 00 00 00 50 51 52
0230 | 53 54 55 56 57 58 00 00 00 00 00 00 00 00 00 00
0240 | 00 00 00 00 00 00 00 00 00 00 00 00 59 5A 5B 5C
0250 | 5D 5E 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0260 | 00 00 00 00 00 00 00 00 00 00 5F 60 61 62 00 00
0270 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0280 | 00 00 00 00 00 00 00 00 00 00 63 64 00 00 00 00
0290 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
02A0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
02B0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
02C0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
02D0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
02E0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
02F0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0300 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0310 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0320 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0330 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0340 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0350 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0360 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0370 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0380 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0390 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
03A0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
03B0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
03C0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
03D0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
03E0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
03F0 | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
material count: 14
country count: 10
block size: 2520
material size: 131074
country size: 3076
low resolution texture count: 7
unaccounted bytes: 1620
```

### CODE REVIEW

File: components/g3d/src/g3d_file.cpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+void G3DFile::ReadBlocks()
+{
+	/*
+	 * The layout of a G3D File is as follows:
+	 *
+	 * - XX byte header, containing:
+	 *         magic header, containing 8 chars "LiOnHeAd"
+	 *
+	 */
```
Comment (bwrsandman) [inline components/g3d/src/g3d_file.cpp]: This is incomplete

File: components/g3d/include/g3d_file.h [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+};
+
+struct imemstream: virtual membuf, std::istream
+{
+	imemstream(char const* base, size_t size)
+		: membuf(base, size)
+		, std::istream(static_cast<std::streambuf*>(this))
+	{
+	}
+};
```
Comment (bwrsandman) [inline components/g3d/include/g3d_file.h]: This is only used internally, move it to cpp file, sync with l3d's version which is more advanced

**Comment by handsomematt** [comment]
excellent stuff so far, I definitely think subcommands would be the way to go.

**Comment by bwrsandman** [comment]
cxxopts does not support subcommands unfortunately

**Pull request merged**